src/server/rdb_save.cc

```diff
@@ -1278,26 +1296,57 @@ error_code RdbSaver::Impl::WriteRecord(io::Bytes src) {
   return ec;
 }
 
-void RdbSaver::Impl::StartSnapshotting(bool stream_journal, const Cancellation* cll,
-                                       EngineShard* shard) {
+void RdbSaver::Impl::PushSnapshotData(Context* cntx, string record) {
+  if (cntx->IsCancelled()) {
+    return;
+  }
+  if (save_mode_ == SaveMode::RDB) {  // Rdb write to channle
+    channel_->Push(record);
+  } else {  // Write directly to socket
+    auto ec = WriteRecord(io::Buffer(record));
+    if (ec) {
+      cntx->ReportError(ec);
+    }
+  }
+}
+
+void RdbSaver::Impl::FinalizeSnapshotWriting() {
+  if (save_mode_ == SaveMode::RDB) {
+    channel_->StartClosing();
+  }
+}
+
+void RdbSaver::Impl::StartSnapshotting(bool stream_journal, Context* cntx, EngineShard* shard) {
   auto& s = GetSnapshot(shard);
   auto& db_slice = namespaces.GetDefaultNamespace().GetDbSlice(shard->shard_id());
-  s = std::make_unique<SliceSnapshot>(&db_slice, &channel_, compression_mode_);
+  auto on_snapshot_finish = std::bind(&RdbSaver::Impl::FinalizeSnapshotWriting, this);
+  auto push_cb = std::bind(&RdbSaver::Impl::PushSnapshotData, this, cntx, std::placeholders::_1);
+
+  s = std::make_unique<SliceSnapshot>(&db_slice, compression_mode_, push_cb, on_snapshot_finish);
 
   const auto allow_flush = (save_mode_ != SaveMode::RDB) ? SliceSnapshot::SnapshotFlush::kAllow
                                                          : SliceSnapshot::SnapshotFlush::kDisallow;
-  s->Start(stream_journal, cll, allow_flush);
+
+  s->Start(stream_journal, cntx->GetCancellation(), allow_flush);
 }
 
 void RdbSaver::Impl::StartIncrementalSnapshotting(Context* cntx, EngineShard* shard,
                                                   LSN start_lsn) {
   auto& db_slice = namespaces.GetDefaultNamespace().GetDbSlice(shard->shard_id());
   auto& s = GetSnapshot(shard);
-  s = std::make_unique<SliceSnapshot>(&db_slice, &channel_, compression_mode_);
+  auto on_finalize_cb = std::bind(&RdbSaver::Impl::FinalizeSnapshotWriting, this);
```

Comment:
it's possible to do like this. not critical but when you pass two lambdas, maybe it's a sign that using an virtual interface is a better pattern here, and make `RdbSaver::Impl` derive from it.